Credit Card Client Information
Cleaned and Sourced from: https://packt.link/wensZ /n
Orginal Dataset from: UCI Machine Learning Repository [http://archive.ics.uci.edu/ml]

In [2]:
import pandas as pd

df = pd.read_excel('./default_of_credit_card_clients__courseware_version_1_21_19.xls')
df.shape

#checking array dimension to match xls file
print('There are X Columns and Y Rows' + str(df.shape))

There are X Columns and Y Rows(30000, 25)


In [3]:
df.info 
df.head(5)

#checking dataframe columns and first 5 entries

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0  798fc410-45c1      20000    2          2         1   24     2      2   
1  8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2  85698822-43f5      90000    2          2         2   34     0      0   
3  0737c11b-be42      50000    2          2         1   37     0      0   
4  3b7f77cc-dbc0      50000    1          2         1   57    -1      0   

   PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0     -1     -1  ...          0          0          0         0       689   
1      0      0  ...       3272       3455       3261         0      1000   
2      0      0  ...      14331      14948      15549      1518      1500   
3      0      0  ...      28314      28959      29547      2000      2019   
4     -1      0  ...      20940      19146      19131      2000     36681   

   PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0         0         0         0         0                           1  
1      1000      1000         0      2000                           1  
2      1000      1000      1000      5000                           0  
3      1200      1100      1069      1000                           0  
4     10000      9000       689       679                           0  

[5 rows x 25 columns]

**Business Problem**

Our client is a credit card company. They have brought us a dataset that includes
some demographics and recent financial data, over the past 6 months, for a sample
of 30,000 of their account holders. This data is at the credit account level; in other
words, there is one row for each account (you should always clarify what the
definition of a row is, in a dataset). Rows are labeled by whether, in the next month
after the 6-month historical data period, an account owner has defaulted, or in other
words, failed to make the minimum payment.

In [4]:
#checking column labels

df.columns

Index(['ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_1',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'default payment next month'],
      dtype='object')

In [5]:
#look for duplicate data entries in dataset by checking amount of unique ID's

print('There are %d unique entries' % df['ID'].nunique())



There are 29687 unique entries


In [6]:
#Count the amount of repeated entries and match them to their ID


id_counts = df['ID'].value_counts()
id_counts.head(10)

id_counts.value_counts()

count
1    29374
2      313
Name: count, dtype: int64

Since there are less than 30,000 unique entries, we needed to count the amount of repeated ID's, and the amount of times an ID is repeated. Occuring once accounts for 29374 duplicate entries, while double duplicated ID account for 313 entries. 

In [7]:
# Create Boolean Mask to locate duplicated ID's

dupe_mask = id_counts == 2
dupe_mask[0:5]


ID
ad23fe5c-7b09    True
1fb3e3e6-a68d    True
89f8f447-fca8    True
7c9b7473-cc2f    True
90330d02-82d9    True
Name: count, dtype: bool

In [10]:
# store duplicate ID in Series, and  onvert it to a list while printing info on list
dupe_ids = id_counts.index[dupe_mask]
dupe_ids = list(dupe_ids)

print("Length of ID's is: %d" %len(dupe_ids))


Length of ID's is: 313


In [11]:
dupe_ids[:5]

['ad23fe5c-7b09',
 '1fb3e3e6-a68d',
 '89f8f447-fca8',
 '7c9b7473-cc2f',
 '90330d02-82d9']

In [12]:
#using the first three ID's in list of duplicates, find all data associated with duplicated ID's to check if, any, differences in data
#are found 
df.loc[df['ID'].isin(dupe_ids[0:3]),:]

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
5033   89f8f447-fca8     320000    2          2         1   32     0      0   
5133   89f8f447-fca8          0    0          0         0    0     0      0   
16727  1fb3e3e6-a68d      80000    1          2         2   33     2      2   
16827  1fb3e3e6-a68d          0    0          0         0    0     0      0   
29685  ad23fe5c-7b09      50000    1          3         1   32     0      0   
29785  ad23fe5c-7b09          0    0          0         0    0     0      0   

       PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
5033       0      0  ...     169371     172868     150827      8000      8000   
5133       0      0  ...          0          0          0         0         0   
16727      0      0  ...      27394      29922      31879         0      2000   
16827      0      0  ...          0          0          0         0         0   
29685      0      0  ...      12882       8131       3983      3000      2871   
29785      0      0  ...          0          0          0         0         0   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
5033       5500      6100      6000      5000                           0  
5133          0         0         0         0                           0  
16727      2000      3000      2600         0                           1  
16827         0         0         0         0                           0  
29685      1000       163      3983      3771                           1  
29785         0         0         0         0                           0  

[6 rows x 25 columns]

It appears duplicate entries in Dataframe contain Null values, having 0 as age makes no sense in our business context, therefore we will delete data entries with 0 values for all other columns accept ID. One reason for repeated values might be faulty join commands on the SQL queries carried out on the Database

In [16]:
#create dataframe that is equal to size of main dataframe holding all account information, fill it with 0 values as place holder
df_zero_mask = df == 0 

#using integer indexing to select all rows, and filter rows by column '1', LIMIT_BAL, then apply all() method along column axis 1

feature_zero_mask = df_zero_mask.iloc[:,1:].all(axis=1)
sum(feature_zero_mask)




315

In [21]:
df_clean_1 = df.loc[~feature_zero_mask,:].copy() #copy df using filter to remove all 0 entries
df_clean_1.shape #check length of Dataframe
df_clean_1.head(10) #check first 10 entries
df_clean_1['ID'].nunique() #print amount of unique entries

29685

In [22]:
#export data to clean CSV with no column info, will add later in next steps 
df_clean_1.to_csv('./df_clean_1.csv', index=False)
